File: src/Algorithms/Lists/Free_Time_Intervals_Simplified/free_time_intervals_simplified.py


# O(n^2) time
# O(n) space
def compare_all_pairs(slots):
    meetings = []

    def is_in_range(slot):
        for index, meeting in enumerate(meetings):
            if ((slot[0] >= meeting[0] and slot[0] <= meeting[1]) or
                (slot[1] >= meeting[0] and slot[1] <= meeting[1]) or
                    (slot[0] <= meeting[0] and slot[1] >= meeting[1])):
                return (index, meeting)

    for slot in slots:
        item = is_in_range(slot)

        if item is not None:
            (index, meeting) = item

            meetings[index] = (
                min(slot[0], meeting[0]),
                max(slot[1], meeting[1])
            )
        else:
            meetings.append(slot)

    return sorted(meetings, key=lambda tup: tup[0])


# O(n*Log(n)) time
# O(1) space
def sort_and_merge(meetings):
    slots = sorted(meetings, key=lambda tup: tup[0])

    i = 0
    while i < len(slots):
        slot = slots[i]

        if i + 1 < len(slots):
            nextSlot = slots[i + 1]

            if (slot[1] >= nextSlot[0]):
                slots[i] = (slot[0], max(slot[1], nextSlot[1]))
                del slots[i+1]
                i -= 1

        i += 1

    return slots

File: src/Algorithms/Lists/Free_Time_Intervals_Simplified/free-time-intervals-simplified.ts
type Tuple = {
  startTime: number;
  endTime: number;
};

// O(n^2) time
// O(n) space
const compareAllPairs = (slots: Tuple[]) => {
  const meetings: Tuple[] = [slots[0]];

  const isInRange = (slot: Tuple) => {
    let meet: Tuple | undefined;
    meetings.forEach((meeting) => {
      if (
        (slot.startTime >= meeting.startTime &&
          slot.startTime <= meeting.endTime) ||
        (slot.endTime >= meeting.startTime &&
          slot.endTime <= meeting.endTime) ||
        (slot.startTime <= meeting.startTime && slot.endTime >= meeting.endTime)
      ) {
        meet = meeting;
      }
    });

    return meet;
  };

  slots.forEach((slot) => {
    const meeting = isInRange(slot);
    if (meeting) {
      meeting.startTime = Math.min(meeting.startTime, slot.startTime);
      meeting.endTime = Math.max(meeting.endTime, slot.endTime);
    } else {
      meetings.push(slot);
    }
  });

  return meetings.sort((a, b) => a.startTime - b.startTime);
};

// O(n*log(n)) time
// O(1) space
const sortAndMerge = (meetings: Tuple[]) => {
  const slots = [...meetings].sort((a, b) => a.startTime - b.startTime);

  for (let i = 0; i < slots.length; i++) {
    const slot = slots[i];
    const next = slots[i + 1];

    if (next && slot.endTime >= next.startTime) {
      slot.endTime = Math.max(slot.endTime, next.endTime);
      slots.splice(i + 1, 1);
      i--;
    }
  }
  return slots;
};

export const solutions = {
  compareAllPairs,
  sortAndMerge,
};
